Refactor fetching of wathola receiver's delivery report using special batch Job (#4460)
knative/eventing: test/upgrade/README.md

entire upgrade/downgrade process. When the upgrade completes, we make sure that
all of those events propagated just once.

To achieve that a [wathola tool](test/upgrade/prober/wathola) was prepared. It
consists of 4 components: _sender_, _forwarder_, _receiver_, and _fetcher_.
_Sender_ is the usual Kubernetes deployment that publishes events to the default

> `broker` with given interval. When it closes (by either `SIGTERM`, or `SIGINT`),

Reviewer: ```suggestion
`broker` with given interval. When it terminates (by either `SIGTERM`, or `SIGINT`),
```

> an `finished` event is generated. _Forwarder_ is a knative serving service that

Reviewer: ```suggestion
a `finished` event is generated. _Forwarder_ is a knative serving service that
```

> scales from zero to receive the requested traffic. _Forwarders_ receive events
> and forwards them to given target. _Receiver_ is an ordinary deployment that
> collects events from multiple forwarders and has an endpoint `/report` that can
> be polled to get the status of sent events. To fetch the report from within the
> cluster _fetcher_ comes in. It's a simple one time job, that will fetch report
> from _receiver_ and print it on stdout as JSON. That enable test client to
> download _fetcher_ logs and parse JSON to get the report.

Reviewer: ```suggestion
scales up from zero to receive the sent events and forward them to given target
which is the _receiver_ in our case. _Receiver_ is an ordinary deployment that
collects events from multiple forwarders and has an endpoint `/report` that can
be polled to get the status of received events. To fetch the report from within the
cluster _fetcher_ comes in. It's a simple one time job, that will fetch the report
from _receiver_ and print it on stdout as JSON. That enables the test client to
download _fetcher_ logs and parse the JSON to get the final report.
```

Author: Hey. In this suggestion you're removing description for forwarder component. Why is that?

Author: Forwarder's are optional component. It's meant to be an additional serving-eventing integration test. Without them sender trigger will beconfigured to point to receiver directly.

Reviewer: @cardil I'm just rephrasing, I guess it wasn't clear that line 61 starts describing the forwarders. It will look like this after my suggestion: 

> _Forwarder_ is a knative serving service that scales up from zero to receive the sent events and forward them to given target which is the _receiver_ in our case.

Author: Ok. Got it.

Diagram below describe the setup:

```
                   K8s cluster                            |     Test machine
                                                          |
 (deploym.)         (ksvc)            (deploym.)          |
+--------+       +-----------+       +----------+         |    +------------+
|        |       |           ++      |          |         |    |            |
| Sender |   +-->| Forwarder ||----->+ Receiver |         |    + TestProber |
|        |   |   |           ||      |          |<---+    |    |            |
+---+----+   |   +------------|      +----------+    |    |    +------------+
    |        |    +-----------+                      |    |
    |        |                                       |    |
    |        |                              +---------+   |
    |     +--+-----+       +---------+      |         |   |
    +----->        |       |         +-+    + Fetcher |   |
          | Broker | < - > | Trigger | |    |         |   |
          |        |       |         | |    +---------+   |
          +--------+       +---------+ |       (job)      |

>            (default)        +----------+                  |

Reviewer 2: Why the receiver and the fetcher are two separate components? 
I'd merge the sender, the receiver, and the fetcher to simplify design and implementation.

Reviewer 2: I imagine a job that sends and receives events and fails when sent events aren't received, in that way the TestProber job is to watch (receiver, sender, fetcher) component final status (Succeeded / Failed).

Author: Why you think it will simplify anything? For me, it looks like it might complicate things.

If done, then something that generates and send events will also expect to receive them. Then report about missing events.  And how you would notify that _something_ that it is time to stop receiving events, and prepare a report?

Multiple responsibilities in one code, it isn't simple, but something that quickly can become spaghetti code.

I think we can discuss here only if we should have _receiver_ and _fetcher_ as separate entities, or merged into one. Merging them with _sender_ is out of the scope of this PR, and we can create an issue to discuss it.

So, about merging _fetcher_ and _receiver_ into a single entity. That will force that receiver will need to have machine-readable logs which we could parse at test runner. And we would need to be able to diversify normal logs (debug, info, warn, error) from report. _fetcher_ is simpler in that case as it is a short-lived container, so we easily can report everything in one go as JSON.

Reviewer: @cardil after a discussion online we can have the future improvements discussion in a separate issue that @pierDipi is thankfully working on
